gpt_neox/datasets.py

```diff
@@ -7,6 +7,86 @@
 import re
 import logging
 from itertools import cycle
+import simdjson as json
+import linecache
+import numpy as np
+
+PAD_TOKEN=50257
+
+"""
+Dataset that gets sequences from a set of sharded jsonl files
+"""
+class JsonShardedDataset(Dataset):
+
+    def __init__(self, shards_filename:str, tokenizer, seq_length:int,
+                 initial_seed:int):
+        # Input parameters.
+        self.shards_filename = shards_filename
+        self.seq_length = seq_length
+        self.initial_seed = initial_seed
+
+        #TODO: Check if using linecache actually helps
+        self.shard_summary = json.loads(linecache.getline(self.shards_filename,1))
+        self.total_files = len(self.shard_summary['file_names'])
+
+        self.data_length = self.shard_summary['total_shards']
+        self.tokenizer = tokenizer
+
+    def __len__(self):
+        return self.data_length
+    
+    def tokenize(self,data):
+        return self.tokenizer(data, max_length=self.seq_length, return_tensors='pt',\
+                padding='max_length', truncation=True)['input_ids']
+
+    def __getitem__(self, idx):
+        try:
+            shard = json.loads(linecache.getline(self.shards_filename,idx+2))
+        except:
+            print("Could not load: ", self.shards_filename, idx+2)
+            return None
+        try:
+            file_idx,shard_line = shard
+            filename = self.shard_summary['file_names'][file_idx-1]   
+        except:
+            print("Could not get the right file name", self.shard_summary['file_names'], file_idx-1)
+            return None
+        try:
+            #if we've already tokenized the input, all we have to do is pad it 
+            if not self.tokenizer:
+                line = linecache.getline(filename,shard_line)
+                line = list(json.loads(line))
+
+                if len(line) == 0:
+                    print("No words")
+                    raise Exception("An example has no words in it.")
+
+                if len(line) < self.seq_length:
+                    line.extend([PAD_TOKEN for _ in range(self.seq_length-len(line))])
```

Comment:
How often does this happen? This should really be very rare or it will start to effect performance... I guess this is another pro for pre-tokenization - we can ensure it never happens at all

Reply:
I added line 62 for debugging more than anything, right now there shouldn't be any examples of length 0.

Line 64/65 is used all the time: I did this to decrease the size of the pre-tokenized files since almost every example has less than 2048 tokens in it and storing padding tokens would greatly increase the size of the file.

Reply:
Timed it out and each run of 64/65 for a single example takes 1-2 ms.